### DISCUSSION TITLE
How to read last n messages?

### DISCUSSION THREAD

**Discussion by copark86** [gabime/spdlog#2118] [Q&A] [answered]
While it might be easy, I really can't find it anywhere.

I have a c++ back-end program that logs everything with spdlog. I would like to send the last 5 messages to the front-end. How do I read last n messages from spdlog?

**Comment by tt4g** [comment, marked as answer]
You can use `spdlog::sinks::ringbuffer_sink`.

**Reply by copark86** [reply]
> You can use `spdlog::sinks::ringbuffer_sink`.

Thanks! This is what I was looking for!

Just one more question. How should I use this? I tried the following but it didn't work.

`
                spdlog::sinks::ringbuffer_sink_mt test(5);
                std::vector<std::string> last_log = test.last_formatted(5);
`
I am using spdlog in the following way.
`spdlog::info("Downloading the serial number");`

**Reply by tt4g** [reply]
Create `spdlog::logger` with `spdlog::sinks::ringbuffer_sink` and call `spdlog::sinks::ringbuffer_sink::last_formatted(size_t)` after logging.

Wiki (Creating loggers with multiple sinks): https://github.com/gabime/spdlog/wiki/2.-Creating-loggers#creating-loggers-with-multiple-sinks

Example:

```cpp
auto ringbuffer_sink = std::make_shared<spdlog::sinks::ringbuffer_sink_mt>(5);

std::vector<spdlog::sink_ptr> sinks;
sinks.push_back(std::make_shared<spdlog::sinks::basic_file_sink_mt>("path/to/file.log"));
sinks.push_back(ringbuffer_sink);

auto logger = std::make_shared<spdlog::logger>("my_logger", sinks.begin(), sinks.end());

logger->info("foo");
logger->info("bar");
logger->info("baz");

// `log_message` contains "foo", "bar" and "baz".
std::vector<std::string> log_message = ringbuffer_sink->last_formatted();

**Reply by copark86** [reply]
> with multiple sinks): https://github.com/gabime/spdlog/wiki/2.-Creating-loggers#creating-loggers-with-multiple-sinks
> 
> Example:

Thanks for the kind explanation. It would be very nice to include your example somewhere.

**Reply by tt4g** [reply]
Added ringbuffer_sink section in Wiki: https://github.com/gabime/spdlog/wiki/4.-Sinks
